packages/data-service/src/data-service.ts

```diff
@@ -666,7 +666,95 @@ export interface DataService {
   configuredKMSProviders(): string[];
 }
 
-export class DataServiceImpl implements DataService {
+const maybePickNs = ([ns]: unknown[]) => {
+  if (typeof ns === 'string') {
+    return { ns };
+  }
+};
+
+const isPromise = <T>(val: any): val is Promise<T> => {
+  return 'then' in val && typeof val.then === 'function';
+};
+
+/**
+ * Translates the error message to something human readable.
+ * @param error - The error.
+ * @returns The error with message translated.
+ */
+const translateErrorMessage = (error: any): Error | { message: string } => {
+  if (typeof error === 'string') {
+    error = { message: error };
+  } else if (!error.message) {
+    error.message = error.err || error.errmsg;
+  }
+  return error;
+};
+
+abstract class WithLogContext {
+  protected abstract _logCtx(): string;
+}
+
+/**
+ * Decorator to do standard op handling that is applied to every public method
+ * of the data service
+ *
+ * - transform error message before throwing
+ * - log method success / failure
+ */
+function op<T extends unknown[], K>(
+  logId: ReturnType<typeof mongoLogId>,
+  pickLogAttrs: (
+    args: T,
+    result?: K extends Promise<infer R> ? R : K
+  ) => unknown | undefined = maybePickNs
+) {
+  return function (
+    target: (this: WithLogContext, ...args: T) => K,
+    context: ClassMethodDecoratorContext<
+      WithLogContext,
+      (this: WithLogContext, ...args: T) => K
+    >
+  ) {
+    const opName = String(context.name);
+    return function (this: WithLogContext, ...args: T): K {
+      const handleResult = (result: any) => {
+        log.info(
+          logId,
+          this._logCtx(),
+          `Running ${opName}`,
+          pickLogAttrs(args, result)
+        );
+        return result;
+      };
+      const handleError = (error: any) => {
+        const err = translateErrorMessage(error);
+        log.error(
+          mongoLogId(1_001_000_058),
+          this._logCtx(),
+          'Failed to perform data service operation',
+          {
+            op: opName,
+            message: err,
+            ...(pickLogAttrs(args) ?? {}),
+          }
+        );
+        throw err;
+      };
+      try {
+        const result = target.call(this, ...args);
+        if (isPromise<K extends Promise<infer R> ? R : never>(result)) {
+          return result.then(handleResult, handleError) as K;
+        } else {
+          return handleResult(result);
+        }
```

Comment:
We return cursors instead of promises in a few places, which unfortunately makes the implementation here a bit less straightforward and deriving correct types tricky, I opted for just using `any` and asserting types instead where getting TS to cooperate was especially tricky.